shell/common/api/electron_api_native_image_mac.mm

```diff
@@ -34,6 +39,49 @@
   return def;
 }
 
+// static
+v8::Local<v8::Promise> NativeImage::CreateThumbnailFromPath(
+    v8::Isolate* isolate,
+    const base::FilePath& path,
+    const gfx::Size& size) {
+  gin_helper::Promise<gfx::Image> promise(isolate);
+  v8::Local<v8::Promise> handle = promise.GetHandle();
+
+  if (size.IsEmpty()) {
+    promise.RejectWithErrorMessage("size must not be empty");
+    return handle;
+  }
+
+  CGSize cg_size = size.ToCGSize();
+  base::ScopedCFTypeRef<CFURLRef> cfurl = base::mac::FilePathToCFURL(path);
+  base::ScopedCFTypeRef<QLThumbnailRef> ql_thumbnail(
+      QLThumbnailCreate(kCFAllocatorDefault, cfurl, cg_size, NULL));
+  __block gin_helper::Promise<gfx::Image> p = std::move(promise);
+  // we do not want to blocking the main thread while waiting for quicklook to
+  // generate the thumbnail
+  QLThumbnailDispatchAsync(
+      ql_thumbnail,
+      dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, /*flags*/ 0), ^{
+        base::ScopedCFTypeRef<CGImageRef> cg_thumbnail(
+            QLThumbnailCopyImage(ql_thumbnail));
+        if (cg_thumbnail) {
+          NSImage* result =
+              [[[NSImage alloc] initWithCGImage:cg_thumbnail
+                                           size:cg_size] autorelease];
+          gfx::Image thumbnail(result);
+          dispatch_async(dispatch_get_main_queue(), ^{
+            p.Resolve(thumbnail);
+          });
+        } else {
+          dispatch_async(dispatch_get_main_queue(), ^{
+            p.RejectWithErrorMessage("unable to retrieve thumbnail preview "
+                                     "image for the given path");
+          });
+        }
+      });
+  return handle;
+}
+
```

Comment:
Do we need per-platform implementations of this?

For example, what if we did something like this, in a non-blocking thread:

```
gfx::ImageSkia original;
electron::util::PopulateImageSkiaRepsFromPath(&original, filename);
auto scaled = gfx::ImageSkiaOperations::CreateResizedImage(original, skia::ImageOperations::RESIZE_GOOD);
gfx::Image image(scaled);
return Create(isolate, image);
```

To be clear, I haven't tested this pseudocode and don't know what the quality of the output is. The reason I think this would be worth trying is:

* Better to maintain one codebase instead of two
* If it works, it would eliminate the "width but not height supported on Windows" wart
* Picks up Linux for free

Reply:
What this API does is create a *new* image from a path rather than resizing it rather than making it a smaller version of a original image

we have different implementations because of the different APIs needed to extract a thumbnail from a path

Reply:
To be clear this constructs a thumbnail / preview / image _representation_ of a file at a given path.  i.e. we aren't resizing an image.  This means you call this API with a path to a word document and the OS will send you back a thumbnail image for that word document.